### PR TITLE
fix assertion merge for degenerate flow

### LINKED ISSUE #47101: Test failed: JIT\\Directed\\cmov\\Float_And_Op_cs_ro\\Float_And_Op_cs_ro.cmd
Job: [runtime-coreclr r2r-extra 20210116.1](https://dev.azure.com/dnceng/public/_build/results?buildId=954601&view=ms.vss-test-web.build-test-results-tab&runId=30120948&paneView=attachments&resultId=107642)

Failed tests:
```
R2R windows x86 Checked jitstress1 @ Windows.10.Amd64.Open
R2R Linux x64 Checked jitstress2_tiered @ Ubuntu.1804.Amd64.Open
R2R windows x86 Checked jitstress2 @ Windows.10.Amd64.Open
R2R windows x86 Checked jitstress2_tiered @ Windows.10.Amd64.Open
R2R windows x86 Checked jitstress1_tiered @ Windows.10.Amd64.Open
R2R Linux x64 Checked jitstress2 @ Ubuntu.1804.Amd64.Open
R2R windows x64 Checked jitstress1_tiered @ Windows.10.Amd64.Open
R2R Linux x64 Checked jitstress1 @ Ubuntu.1804.Amd64.Open
R2R windows x64 Checked jitstress2_tiered @ Windows.10.Amd64.Open
R2R windows x64 Checked jitstress1 @ Windows.10.Amd64.Open
R2R windows x64 Checked jitstress2 @ Windows.10.Amd64.Open
R2R Linux x64 Checked jitstress1_tiered @ Ubuntu.1804.Amd64.Open
R2R Linux arm64 Checked jitstress2 @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-a45aeeb-20190620155855
R2R Linux arm64 Checked jitstress1 @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-a45aeeb-20190620155855
R2R Linux arm64 Checked jitstress2_tiered @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-a45aeeb-20190620155855
R2R Linux arm64 Checked jitstress1_tiered @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-a45aeeb-20190620155855
R2R Linux arm64 Checked jitstress2_tiered @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-a45aeeb-20190620155855
R2R windows arm64 Checked jitstress1_tiered @ Windows.10.Arm64v8.Open
R2R windows arm64 Checked jitstress1 @ Windows.10.Arm64v8.Open
R2R Linux arm Checked jitstress2_tiered @ (Ubuntu.1804.Arm32.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm32v7-bfcd90a-20200121150440
R2R Linux arm Checked jitstress1 @ (Ubuntu.1804.Arm32.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm32v7-bfcd90a-20200121150440
R2R Linux arm Checked jitstress1_tiered @ (Ubuntu.1804.Arm32.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm32v7-bfcd90a-20200121150440
R2R Linux arm Checked jitstress2 @ (Ubuntu.1804.Arm32.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm32v7-bfcd90a-20200121150440
R2R windows arm64 Checked jitstress2 @ Windows.10.Arm64v8.Open
        - JIT\\Directed\\cmov\\Float_And_Op_cs_ro\\Float_And_Op_cs_ro.cmd
```

Error message:
```
Return code:      1
Raw output file:      C:\h\w\B95209CD\w\B6530A21\e\JIT\Directed\Reports\JIT.Directed\cmov\Float_And_Op_cs_ro\Float_And_Op_cs_ro.output.txt
Raw output:
BEGIN EXECUTION
        1 file(s) copied.
" "C:\h\w\B95209CD\p\crossgen.exe"  /Platform_Assemblies_Paths C:\h\w\B95209CD\p;C:\h\w\B95209CD\w\B6530A21\e\JIT\Directed\cmov\Float_And_Op_cs_ro /in C:\h\w\B95209CD\w\B6530A21\e\JIT\Directed\cmov\Float_And_Op_cs_ro\Float_And_Op_cs_ro.org /out C:\h\w\B95209CD\w\B6530A21\e\JIT\Directed\cmov\Float_And_Op_cs_ro\\Float_And_Op_cs_ro.dll"
Microsoft (R) CoreCLR Native Image Generator - Version 6.0.0-ci
Copyright (c) Microsoft Corporation.  All rights reserved.

Native image C:\h\w\B95209CD\w\B6530A21\e\JIT\Directed\cmov\Float_And_Op_cs_ro\\Float_And_Op_cs_ro.dll generated successfully.
 "C:\h\w\B95209CD\p\corerun.exe" Float_And_Op_cs_ro.dll 
FAILED
Expected: 100
Actual: 1
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=C:\h\w\B95209CD\p
> C:\h\w\B95209CD\w\B6530A21\e\JIT\Directed\cmov\Float_And_Op_cs_ro\Float_And_Op_cs_ro.cmd

stack trace:
   at JIT_Directed._cmov_Float_And_Op_cs_ro_Float_And_Op_cs_ro_._cmov_Float_And_Op_cs_ro_Float_And_Op_cs_ro_cmd()
```

### PR DESCRIPTION
**Pull request by kunalspathak**
Sometimes, we have degenerate flow involving 2 blocks A and B such that `A->jumpDst == B` as well as `A->bbNext == B`. In such case we should merge the assertions of out edges from both flows into the B's in edge. Eventually we want to avoid generating such flows in redundant branch optimization. See https://github.com/dotnet/runtime/issues/48609.

Thanks Andy for help identify the root cause.

Fixes:  #47101

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib , @AndyAyersMS

**Comment by kunalspathak** [comment]
https://dev.azure.com/dnceng/public/_build/results?buildId=1006551&view=results is green again with this change.

**Review by briansull — APPROVED** [review]
Looks Good

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -4648,14 +4648,28 @@ class AssertionPropFlowCallback
...
-        ASSERT_TP pAssertionOut = ((predBlock->bbJumpKind == BBJ_COND) && (predBlock->bbJumpDest == block))
-                                      ? mJumpDestOut[predBlock->bbNum]
-                                      : predBlock->bbAssertionOut;
+        ASSERT_TP pAssertionOut;
+
+        if (predBlock->bbJumpKind == BBJ_COND && (predBlock->bbJumpDest == block))
+        {
+            pAssertionOut = mJumpDestOut[predBlock->bbNum];
+
+            if (dupCount > 1)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/assertionprop.cpp:4659]: This deserves a comment and some extra jit dump output, since it is an unusual case.

**Review by AndyAyersMS — COMMENTED** [review]
Did you get a chance to look at other merges to see if any of them might be impacted by the same problem?

File: src/coreclr/jit/assertionprop.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/assertionprop.cpp:4659]: Agree. Added.

**Comment by kunalspathak** [comment]
> Did you get a chance to look at other merges to see if any of them might be impacted by the same problem?

Are you asking about CSE? Apart from AssertionProp, that's the only one that uses `DataFlow` and perform `Merge`. For CSE, I didn't see logic where we chose which edge to merge so I think it should be fine.

If you meant something else, can you please clarify other phases that does similar thing and I will double check.

**Comment by AndyAyersMS** [comment]
Did you get a chance to run diffs for this? Curious if it popped up anywhere else, and why we didn't notice it, if so.

**Comment by kunalspathak** [comment]
> Did you get a chance to run diffs for this? Curious if it popped up anywhere else, and why we didn't notice it, if so.

Yes, I ran libraries.pmi, tests and benchmarks superpmi collection and found no asmdiff. The original issue was in R2R with jitstress flag. While investigating, I tried to repro using JIT, reduce jitstress flags, etc. but none of them helped in reproing so I believe that it took a unique combination of things to get to this, so "no asm diff" don't surprise me.

**Pull request merged**